File: fons/codegen/zig/norma/tabula.ts
/**
 * Tabula Method Registry - Zig translations for Latin map methods
 *
 * COMPILER PHASE
 * ==============
 * codegen (Zig target)
 *
 * ARCHITECTURE
 * ============
 * This module defines Zig translations for tabula<K,V> (HashMap) methods.
 * Zig's HashMap requires an allocator for most operations.
 *
 * ZIG SPECIFICS
 * =============
 * - tabula<textus, V> maps to std.StringHashMap(V)
 * - tabula<K, V> maps to std.AutoHashMap(K, V)
 * - Operations need allocator: map.put(alloc, key, value)
 * - get() returns optional: map.get(key) orelse default
 *
 * INPUT/OUTPUT CONTRACT
 * =====================
 * INPUT:  Latin method name from CallExpression
 * OUTPUT: Zig code string
 * ERRORS: Returns undefined if method name not recognized
 */

// =============================================================================
// TYPES
// =============================================================================

/**
 * Generator function type for Zig collection methods.
 * WHY: The curator parameter allows methods to use the correct allocator.
 */
export type ZigGenerator = (obj: string, args: string, curator: string) => string;

export interface TabulaMethod {
    latin: string;
    mutates: boolean;
    zig: string | ZigGenerator;
}

// =============================================================================
// METHOD REGISTRY
// =============================================================================

export const TABULA_METHODS: Record<string, TabulaMethod> = {
    // -------------------------------------------------------------------------
    // CORE OPERATIONS
    // -------------------------------------------------------------------------

    /** Set key-value pair (mutates, needs allocator) */
    pone: {
        latin: 'pone',
        mutates: true,
        // WHY: put() can fail on OOM
        // args format: "key, value" - need to split
        zig: (obj, args, curator) => {
            const match = args.match(/^(.+?),\s*(.+)$/);
            if (match) {
                return `${obj}.put(${curator}, ${match[1]}, ${match[2]}) catch @panic("OOM")`;
            }
            return `@compileError("pone requires two arguments: key, value")`;
        },
    },

    /** Get value by key (returns optional) */
    accipe: {
        latin: 'accipe',
        mutates: false,
        zig: (obj, args) => `${obj}.get(${args})`,
    },

    /** Check if key exists */
    habet: {
        latin: 'habet',
        mutates: false,
        zig: (obj, args) => `${obj}.contains(${args})`,
    },

    /** Delete key (mutates) */
    dele: {
        latin: 'dele',
        mutates: true,
        // WHY: remove() returns bool indicating if key existed
        zig: (obj, args) => `_ = ${obj}.remove(${args})`,
    },

    /** Get size */
    longitudo: {
        latin: 'longitudo',
        mutates: false,
        zig: (obj, _args) => `${obj}.count()`,
    },

    /** Check if empty */
    vacua: {
        latin: 'vacua',
        mutates: false,
        zig: (obj, _args) => `(${obj}.count() == 0)`,
    },

    /** Clear all entries (mutates) */
    purga: {
        latin: 'purga',
        mutates: true,
        zig: (obj, _args) => `${obj}.clearRetainingCapacity()`,
    },

    // -------------------------------------------------------------------------
    // ITERATION
    // -------------------------------------------------------------------------

    /** Get keys iterator */
    claves: {
        latin: 'claves',
        mutates: false,
        // WHY: keyIterator() returns iterator over keys
        zig: (obj, _args) => `${obj}.keyIterator()`,
    },

    /** Get values iterator */
    valores: {
        latin: 'valores',
        mutates: false,
        zig: (obj, _args) => `${obj}.valueIterator()`,
    },

    /** Get entries iterator */
    paria: {
        latin: 'paria',
        mutates: false,
        // WHY: iterator() returns key-value pairs
        zig: (obj, _args) => `${obj}.iterator()`,
    },

    // -------------------------------------------------------------------------
    // EXTENDED OPERATIONS - NOT IMPLEMENTED
    // -------------------------------------------------------------------------

    /** Get value or return default */
    accipeAut: {
        latin: 'accipeAut',
        mutates: false,
        // WHY: Use orelse for default value
        zig: (obj, args) => {
            const match = args.match(/^(.+?),\s*(.+)$/);
            if (match) {
                return `(${obj}.get(${match[1]}) orelse ${match[2]})`;
            }
            return `${obj}.get(${args})`;
        },
    },

    /** Keep only specified keys - NOT IMPLEMENTED */
    selige: {
        latin: 'selige',
        mutates: false,
        zig: () => `@compileError("selige not implemented for Zig - use explicit loop")`,
    },

    /** Remove specified keys - NOT IMPLEMENTED */
    omitte: {
        latin: 'omitte',
        mutates: false,
        zig: () => `@compileError("omitte not implemented for Zig - use explicit loop")`,
    },

    /** Merge maps - NOT IMPLEMENTED */
    confla: {
        latin: 'confla',
        mutates: false,
        zig: () => `@compileError("confla not implemented for Zig - use explicit loop")`,
    },

    /** Swap keys and values - NOT IMPLEMENTED */
    inversa: {
        latin: 'inversa',
        mutates: false,
        zig: () => `@compileError("inversa not implemented for Zig - use explicit loop")`,
    },

    /** Transform values - NOT IMPLEMENTED */
    mappaValores: {
        latin: 'mappaValores',
        mutates: false,
        zig: () => `@compileError("mappaValores not implemented for Zig - use explicit loop")`,
    },

    /** Transform keys - NOT IMPLEMENTED */
    mappaClaves: {
        latin: 'mappaClaves',
        mutates: false,
        zig: () => `@compileError("mappaClaves not implemented for Zig - use explicit loop")`,
    },

    // -------------------------------------------------------------------------
    // CONVERSIONS - NOT IMPLEMENTED
    // -------------------------------------------------------------------------

    /** Convert to list of entries */
    inLista: {
        latin: 'inLista',
        mutates: false,
        zig: () => `@compileError("inLista not implemented for Zig - iterate with ex...pro")`,
    },

    /** Convert to object */
    inObjectum: {
        latin: 'inObjectum',
        mutates: false,
        zig: () => `@compileError("inObjectum not implemented for Zig - Zig has no object type")`,
    },
};

// =============================================================================
// LOOKUP FUNCTIONS
// =============================================================================

export function getTabulaMethod(name: string): TabulaMethod | undefined {
    return TABULA_METHODS[name];
}

export function isTabulaMethod(name: string): boolean {
    return name in TABULA_METHODS;
}

File: subsidia/zig/tabula.zig
// Tabula - HashMap wrapper for Faber
// Wraps std.StringHashMap/AutoHashMap with Latin method names
//
// WHY: No stored allocator. Methods that need allocation accept it as parameter.
// This ensures cura...fit blocks properly thread the current allocator.

const std = @import("std");

/// Tabula(K, V) is a hash map with Latin-named methods.
/// For string keys, use TabulaTextus(V) which wraps StringHashMap.
/// For other keys, use TabulaAuto(K, V) which wraps AutoHashMap.
///
/// WHY: Zig's HashMap API requires allocator for put() but not get().
/// We follow that pattern: only mutating/growing operations need allocator.
pub fn Tabula(comptime K: type, comptime V: type) type {
    // WHY: String keys use StringHashMap for better performance
    if (K == []const u8) {
        return TabulaTextus(V);
    }
    return TabulaAuto(K, V);
}

/// Tabula for string keys - wraps std.StringHashMap
pub fn TabulaTextus(comptime V: type) type {
    return struct {
        map: std.StringHashMap(V),

        const Self = @This();

        // =====================================================================
        // CONSTRUCTION
        // =====================================================================

        /// Create an empty tabula.
        pub fn init(alloc: std.mem.Allocator) Self {
            return .{ .map = std.StringHashMap(V).init(alloc) };
        }

        /// Free the tabula's memory.
        pub fn deinit(self: *Self) void {
            self.map.deinit();
        }

        // =====================================================================
        // CORE OPERATIONS
        // =====================================================================

        /// Set key-value pair (mutates, needs allocator for growth).
        pub fn pone(self: *Self, alloc: std.mem.Allocator, key: []const u8, value: V) void {
            _ = alloc; // StringHashMap tracks its allocator internally
            self.map.put(key, value) catch @panic("OOM");
        }

        /// Get value by key (returns optional).
        pub fn accipe(self: Self, key: []const u8) ?V {
            return self.map.get(key);
        }

        /// Get value or return default.
        pub fn accipeAut(self: Self, key: []const u8, default: V) V {
            return self.map.get(key) orelse default;
        }

        /// Check if key exists.
        pub fn habet(self: Self, key: []const u8) bool {
            return self.map.contains(key);
        }

        /// Delete key (mutates).
        pub fn dele(self: *Self, key: []const u8) bool {
            return self.map.remove(key);
        }

        /// Get size.
        pub fn longitudo(self: Self) usize {
            return self.map.count();
        }

        /// Check if empty.
        pub fn vacua(self: Self) bool {
            return self.map.count() == 0;
        }

        /// Clear all entries (mutates).
        pub fn purga(self: *Self) void {
            self.map.clearRetainingCapacity();
        }

        // =====================================================================
        // ITERATION
        // =====================================================================

        /// Get keys iterator.
        pub fn claves(self: *Self) std.StringHashMap(V).KeyIterator {
            return self.map.keyIterator();
        }

        /// Get values iterator.
        pub fn valores(self: *Self) std.StringHashMap(V).ValueIterator {
            return self.map.valueIterator();
        }

        /// Get entries iterator.
        pub fn paria(self: *Self) std.StringHashMap(V).Iterator {
            return self.map.iterator();
        }

        // =====================================================================
        // TRANSFORMATION
        // =====================================================================

        /// Merge another tabula into this one (mutates).
        pub fn confla(self: *Self, other: *Self) void {
            var iter = other.map.iterator();
            while (iter.next()) |entry| {
                self.map.put(entry.key_ptr.*, entry.value_ptr.*) catch @panic("OOM");
            }
        }

        /// Convert to list of key-value tuples.
        pub fn inLista(self: *Self, alloc: std.mem.Allocator) []struct { []const u8, V } {
            const Pair = struct { []const u8, V };
            var result = std.ArrayList(Pair).init(alloc);
            var iter = self.map.iterator();
            while (iter.next()) |entry| {
                result.append(.{ entry.key_ptr.*, entry.value_ptr.* }) catch @panic("OOM");
            }
            return result.toOwnedSlice() catch @panic("OOM");
        }
    };
}

/// Tabula for non-string keys - wraps std.AutoHashMap
pub fn TabulaAuto(comptime K: type, comptime V: type) type {
    return struct {
        map: std.AutoHashMap(K, V),

        const Self = @This();

        // =====================================================================
        // CONSTRUCTION
        // =====================================================================

        /// Create an empty tabula.
        pub fn init(alloc: std.mem.Allocator) Self {
            return .{ .map = std.AutoHashMap(K, V).init(alloc) };
        }

        /// Free the tabula's memory.
        pub fn deinit(self: *Self) void {
            self.map.deinit();
        }

        // =====================================================================
        // CORE OPERATIONS
        // =====================================================================

        /// Set key-value pair (mutates, needs allocator for growth).
        pub fn pone(self: *Self, alloc: std.mem.Allocator, key: K, value: V) void {
            _ = alloc; // AutoHashMap tracks its allocator internally
            self.map.put(key, value) catch @panic("OOM");
        }

        /// Get value by key (returns optional).
        pub fn accipe(self: Self, key: K) ?V {
            return self.map.get(key);
        }

        /// Get value or return default.
        pub fn accipeAut(self: Self, key: K, default: V) V {
            return self.map.get(key) orelse default;
        }

        /// Check if key exists.
        pub fn habet(self: Self, key: K) bool {
            return self.map.contains(key);
        }

        /// Delete key (mutates).
        pub fn dele(self: *Self, key: K) bool {
            return self.map.remove(key);
        }

        /// Get size.
        pub fn longitudo(self: Self) usize {
            return self.map.count();
        }

        /// Check if empty.
        pub fn vacua(self: Self) bool {
            return self.map.count() == 0;
        }

        /// Clear all entries (mutates).
        pub fn purga(self: *Self) void {
            self.map.clearRetainingCapacity();
        }

        // =====================================================================
        // ITERATION
        // =====================================================================

        /// Get keys iterator.
        pub fn claves(self: *Self) std.AutoHashMap(K, V).KeyIterator {
            return self.map.keyIterator();
        }

        /// Get values iterator.
        pub fn valores(self: *Self) std.AutoHashMap(K, V).ValueIterator {
            return self.map.valueIterator();
        }

        /// Get entries iterator.
        pub fn paria(self: *Self) std.AutoHashMap(K, V).Iterator {
            return self.map.iterator();
        }

        // =====================================================================
        // TRANSFORMATION
        // =====================================================================

        /// Merge another tabula into this one (mutates).
        pub fn confla(self: *Self, other: *Self) void {
            var iter = other.map.iterator();
            while (iter.next()) |entry| {
                self.map.put(entry.key_ptr.*, entry.value_ptr.*) catch @panic("OOM");
            }
        }

        /// Convert to list of key-value tuples.
        pub fn inLista(self: *Self, alloc: std.mem.Allocator) []struct { K, V } {
            const Pair = struct { K, V };
            var result = std.ArrayList(Pair).init(alloc);
            var iter = self.map.iterator();
            while (iter.next()) |entry| {
                result.append(.{ entry.key_ptr.*, entry.value_ptr.* }) catch @panic("OOM");
            }
            return result.toOwnedSlice() catch @panic("OOM");
        }
    };
}
